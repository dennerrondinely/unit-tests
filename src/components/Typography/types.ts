export type Component = 'h1' | 'h2' | 'caption' | 'span' | 'p';

export type Variant =
  | 'heading-1'
  | 'heading-2'
  | 'subtitle'
  | 'body'
  | 'caption';

export enum VariantType {
  'heading-1' = 'h1',
  'heading-2' = 'h1',
  subtitle = 'h2',
  'body' = 'p',
  'caption' = 'caption'
}
