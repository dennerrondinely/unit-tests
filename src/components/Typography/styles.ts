import styled, { css, DefaultTheme } from 'styled-components';
import media from 'styled-media-query';

import { TypographyProps } from '.';

type WrapperProps = { theme: DefaultTheme } & Omit<TypographyProps, 'children'>;

const includeNowrap = css`
  white-space: nowrap;
  width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const getVariant = {
  'heading-1': ({
    theme,
    nowrap,
    align = 'center',
    color = 'blue-dark'
  }: WrapperProps) => css`
    color: ${theme.colors[color]};
    font-style: normal;
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.xxlarge};
    line-height: 3.9rem;
    letter-spacing: -1.5px;
    text-align: ${align};
    ${nowrap && includeNowrap}

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.huge};
      line-height: 6rem;
    `}
  `,
  'heading-2': ({
    theme,
    nowrap,
    align = 'start',
    color = 'blue-dark'
  }: WrapperProps) => css`
    color: ${theme.colors[color]};
    font-style: normal;
    font-weight: ${theme.font.bold};
    font-size: ${theme.font.sizes.xlarge};
    line-height: 3.8rem;
    text-align: ${align};
    ${nowrap && includeNowrap}

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.huge};
      line-height: 6.5rem;
    `}
  `,
  body: ({
    theme,
    nowrap,
    align = 'start',
    color = 'gray-dark'
  }: WrapperProps) => css`
    font-style: normal;
    font-weight: ${theme.font.normal};
    font-size: ${theme.font.sizes.small};
    line-height: 2rem;
    color: ${theme.colors[color]};
    text-align: ${align};
    ${nowrap && includeNowrap}
  `,
  subtitle: ({
    theme,
    nowrap,
    align = 'center',
    color = 'gray-medium'
  }: WrapperProps) =>
    css`
      font-style: normal;
      font-weight: ${theme.font.bold};
      font-size: ${theme.font.sizes.medium};
      line-height: 2.4rem;
      text-align: ${align};
      color: ${theme.colors[color]};
      ${nowrap && includeNowrap}
    `,
  caption: ({
    theme,
    nowrap,
    align = 'start',
    color = 'gray-dark'
  }: WrapperProps) =>
    css`
      font-style: normal;
      font-weight: ${theme.font.normal};
      font-size: ${theme.font.sizes.xsmall};
      line-height: 2rem;
      text-align: ${align};
      color: ${theme.colors[color]};
      ${nowrap && includeNowrap}
    `
};

export const h1 = styled.h1<WrapperProps>`
  ${({ variant = 'heading-1', ...props }: WrapperProps) =>
    getVariant[variant](props)}
`;
export const h2 = styled.h2<WrapperProps>`
  ${({ variant = 'subtitle', ...props }: WrapperProps) =>
    getVariant[variant](props)}
`;
export const span = styled.span<WrapperProps>`
  ${({ variant = 'body', ...props }: WrapperProps) =>
    getVariant[variant](props)}
`;
export const p = styled.p<WrapperProps>`
  ${({ variant = 'body', ...props }) => getVariant[variant](props)}
`;
export const caption = styled.caption<WrapperProps>`
  ${({ variant = 'caption', ...props }) => getVariant[variant](props)}
`;
