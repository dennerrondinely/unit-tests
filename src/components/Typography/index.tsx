import React, { ReactNode } from 'react';

import { ColorsType } from 'styles/colors';

import { Component, Variant, VariantType } from './types';

import * as S from './styles';

export type TypographyProps = {
  variant?: Variant;
  component?: Component;
  color?: ColorsType;
  nowrap?: boolean;
  align?: 'start' | 'center' | 'end';
  children?: ReactNode | string;
  role?: string;
};

const Typography = ({
  children,
  variant = 'body',
  component,
  align,
  color,
  nowrap,
  role
}: TypographyProps) => {
  const Component = S[component ?? VariantType[variant]];

  return (
    <Component
      variant={variant}
      align={align}
      color={color}
      nowrap={nowrap}
      role={role}
    >
      {children}
    </Component>
  );
};

export default Typography;
