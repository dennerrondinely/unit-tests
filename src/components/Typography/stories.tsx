import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import Typography, { TypographyProps } from '.';

export default {
  title: 'Typography',
  component: Typography,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta;

export const Default: Story<TypographyProps> = (args) => (
  <Typography {...args} />
);

Default.args = {
  children: 'Consultas',
  color: 'gray-dark'
};
