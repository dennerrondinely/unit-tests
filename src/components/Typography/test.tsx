import React from 'react';

import Typography from './index';
import { render, screen } from 'utils/test-utils';

describe('<Typography />', () => {
  it('should render the heading h1', () => {
    const { container } = render(
      <Typography component="h1">Typography</Typography>
    );

    expect(
      screen.getByRole('heading', { name: /Typography/i })
    ).toBeInTheDocument();

    expect(screen.getByRole('heading', { name: /Typography/i })).toContainHTML(
      'h1'
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the heading h2', () => {
    const { container } = render(
      <Typography component="h2">Typography</Typography>
    );

    expect(
      screen.getByRole('heading', { name: /Typography/i })
    ).toBeInTheDocument();

    expect(screen.getByRole('heading', { name: /Typography/i })).toContainHTML(
      'h2'
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the span', () => {
    const { container } = render(
      <Typography component="span">Typography</Typography>
    );

    expect(screen.getByText(/Typography/i)).toBeInTheDocument();

    expect(screen.getByText(/Typography/i)).toContainHTML('span');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the heading p', () => {
    const { container } = render(
      <Typography component="p">Typography</Typography>
    );

    expect(screen.getByText(/Typography/i)).toBeInTheDocument();

    expect(screen.getByText(/Typography/i)).toContainHTML('p');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the color', () => {
    render(<Typography color="blue">Typography</Typography>);

    expect(screen.getByText(/Typography/i)).toBeInTheDocument();

    expect(screen.getByText(/Typography/i)).toHaveStyle({
      color: '#2E50D4'
    });
  });
});
