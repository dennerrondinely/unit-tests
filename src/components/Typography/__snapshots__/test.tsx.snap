// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> should render the heading h1 1`] = `
.c0 {
  font-style: normal;
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 2rem;
  color: #575453;
  text-align: start;
}

<h1
  class="c0"
>
  Typography
</h1>
`;

exports[`<Typography /> should render the heading h2 1`] = `
.c0 {
  font-style: normal;
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 2rem;
  color: #575453;
  text-align: start;
}

<h2
  class="c0"
>
  Typography
</h2>
`;

exports[`<Typography /> should render the heading p 1`] = `
.c0 {
  font-style: normal;
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 2rem;
  color: #575453;
  text-align: start;
}

<p
  class="c0"
>
  Typography
</p>
`;

exports[`<Typography /> should render the span 1`] = `
.c0 {
  font-style: normal;
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 2rem;
  color: #575453;
  text-align: start;
}

<span
  class="c0"
>
  Typography
</span>
`;
